File: mobile/src/main/kotlin/de/scout/fireplace/settings/NumericRangeView.java
package de.scout.fireplace.settings;

import android.content.Context;
import android.content.res.TypedArray;
import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputLayout;
import android.text.InputFilter;
import android.util.AttributeSet;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.TextView;
import com.google.firebase.crash.FirebaseCrash;
import de.scout.fireplace.R;

public class NumericRangeView extends FrameLayout {

  public NumericRangeView(@NonNull Context context) {
    super(context);
    init(context, null, 0);
  }

  public NumericRangeView(@NonNull Context context, @Nullable AttributeSet attrs) {
    super(context, attrs);
    init(context, attrs, 0);
  }

  public NumericRangeView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
    init(context, attrs, defStyleAttr);
  }

  private void init(Context context, AttributeSet attrs, int defStyleAttr) {
    View.inflate(context, R.layout.numeric_range_view, this);

    TypedArray attributes = context.obtainStyledAttributes(attrs, R.styleable.NumericRangeView);
    try {
      getTextView(R.id.label).setText(attributes.getString(R.styleable.NumericRangeView_label));

      getInputLayout(R.id.minimum).setHint(attributes.getString(R.styleable.NumericRangeView_minHint));
      getInputLayout(R.id.minimum).setHint(attributes.getString(R.styleable.NumericRangeView_minValue));

      getInputLayout(R.id.maximum).setHint(attributes.getString(R.styleable.NumericRangeView_maxHint));
      getInputLayout(R.id.maximum).setHint(attributes.getString(R.styleable.NumericRangeView_maxValue));
    } finally {
      attributes.recycle();
    }
  }

  private TextView getTextView(@IdRes int resId) {
    return findViewById(resId);
  }

  private TextInputLayout getInputLayout(@IdRes int resId) {
    return (TextInputLayout) findViewById(resId);
  }

  public void setMinHint(String minHint) {
    getInputLayout(R.id.minimum).setHint(minHint);
  }

  public void setMinValue(int minValue) {
    getInputLayout(R.id.minimum).getEditText().setText(String.valueOf(minValue));
  }

  public void setMaxHint(String maxHint) {
    getInputLayout(R.id.maximum).setHint(maxHint);
  }

  public void setMaxValue(int maxValue) {
    getInputLayout(R.id.maximum).getEditText().setText(String.valueOf(maxValue));
  }

  public void setFilter(InputFilter... filters) {
    getInputLayout(R.id.minimum).getEditText().setFilters(filters);
    getInputLayout(R.id.maximum).getEditText().setFilters(filters);
  }

  public void addOnMinChangeListener(OnMinChangeListener listener) {
    getInputLayout(R.id.minimum).getEditText().addTextChangedListener(new SimpleTextWatcher(value -> listener.onMinChange(parseFloat(value))));
  }

  public void addOnMaxChangeListener(OnMaxChangeListener listener) {
    getInputLayout(R.id.maximum).getEditText().addTextChangedListener(new SimpleTextWatcher(value -> listener.onMaxChange(parseFloat(value))));
  }

  private static float parseFloat(String string) {
    if (string == null || string.isEmpty()) {
      return 0;
    }

    try {
      return Float.parseFloat(string.replaceAll("[^\\d]", ""));
    } catch (NumberFormatException exception) {
      FirebaseCrash.report(exception);
      return 0F;
    }
  }

  interface OnMinChangeListener {

    void onMinChange(float value);
  }

  interface OnMaxChangeListener {

    void onMaxChange(float value);
  }
}

File: feature/src/main/kotlin/de/scout/fireplace/feature/settings/NumericRangeView.kt
package de.scout.fireplace.settings

import android.content.Context
import android.support.annotation.IdRes
import android.support.design.widget.TextInputLayout
import android.text.InputFilter
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout
import android.widget.TextView
import com.google.firebase.crash.FirebaseCrash
import de.scout.fireplace.feature.R

internal class NumericRangeView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : FrameLayout(context, attrs, defStyleAttr) {

  init {
    View.inflate(context, R.layout.numeric_range_view, this)

    val attributes = context.obtainStyledAttributes(attrs, R.styleable.NumericRangeView)
    try {
      getTextView(R.id.label).text = attributes.getString(R.styleable.NumericRangeView_label)

      getInputLayout(R.id.minimum).hint = attributes.getString(R.styleable.NumericRangeView_minHint)
      getInputLayout(R.id.minimum).hint = attributes.getString(R.styleable.NumericRangeView_minValue)

      getInputLayout(R.id.maximum).hint = attributes.getString(R.styleable.NumericRangeView_maxHint)
      getInputLayout(R.id.maximum).hint = attributes.getString(R.styleable.NumericRangeView_maxValue)
    } finally {
      attributes.recycle()
    }
  }

  private fun getTextView(@IdRes resId: Int): TextView {
    return findViewById(resId)
  }

  private fun getInputLayout(@IdRes resId: Int): TextInputLayout {
    return findViewById<View>(resId) as TextInputLayout
  }

  fun setMinHint(minHint: String) {
    getInputLayout(R.id.minimum).hint = minHint
  }

  fun setMinValue(minValue: Int) {
    getInputLayout(R.id.minimum).editText!!.setText(minValue.toString())
  }

  fun setMaxHint(maxHint: String) {
    getInputLayout(R.id.maximum).hint = maxHint
  }

  fun setMaxValue(maxValue: Int) {
    getInputLayout(R.id.maximum).editText!!.setText(maxValue.toString())
  }

  fun setFilter(vararg filters: InputFilter) {
    getInputLayout(R.id.minimum).editText!!.filters = filters
    getInputLayout(R.id.maximum).editText!!.filters = filters
  }

  fun addOnMinChangeListener(listener: (Float) -> Unit) {
    getInputLayout(R.id.minimum).editText!!.addTextChangedListener(SimpleTextWatcher { value -> listener(parseFloat(value)) })
  }

  fun addOnMaxChangeListener(listener: (Float) -> Unit) {
    getInputLayout(R.id.maximum).editText!!.addTextChangedListener(SimpleTextWatcher { value -> listener(parseFloat(value)) })
  }

  private fun parseFloat(string: String?): Float {
    if (string == null || string.isEmpty()) {
      return 0f
    }

    try {
      return java.lang.Float.parseFloat(string.replace("[^\\d]".toRegex(), ""))
    } catch (exception: NumberFormatException) {
      FirebaseCrash.report(exception)
      return 0f
    }

  }

  internal interface OnMinChangeListener {

    fun onMinChange(value: Float)
  }

  internal interface OnMaxChangeListener {

    fun onMaxChange(value: Float)
  }
}
